File: lab2-functionalprimer/javascript/func_as_arg.js
// Create a function to generate a list of numbers from a to b
function arrFunc(a, b) {
    let arr = [];
    for (let i = a; i <= b; i++) {
        arr.push(i);
    }
    return arr;
}

// Define applicatorFunc to compute sum or average
function applicatorFunc(inpFunc, a, b, s) {
    const arr = inpFunc(a, b);
    let sum = arr.reduce((accumulator, currentValue) => accumulator + currentValue, 0);

    return s === 's' ? sum : sum / arr.length;
}

// Call applicatorFunc with range 1 to 10 and choice 's' (sum) or 'a' (average)
console.log(applicatorFunc(arrFunc, 1, 10, 's')); // Sum
console.log(applicatorFunc(arrFunc, 1, 10, 'a')); // Average

File: lab2-functionalprimer/haskell/func_as_arg.hs
-- Create inpFunc
inpFunc :: Int -> Int -> [Int]
inpFunc a b = [a..b]

-- Define applicatorFunc
applicatorFunc :: (Int -> Int -> [Int]) -> Int -> Int -> Char -> Double
applicatorFunc inpFunc a b s =
    if s == 's' 
        then fromIntegral (sum (inpFunc a b)) 
        else fromIntegral (sum (inpFunc a b)) / fromIntegral (b - a + 1)

main :: IO ()
main = do
    let result = applicatorFunc inpFunc 1 10 's' -- Call applicatorFunc with inpFunc, integers, and char as args
    putStrLn ("sum = " ++ show result)
